xml_converter/src/xml_converter.cpp

```diff
@@ -149,20 +152,30 @@ void write_protobuf_file(string proto_filepath, map<string, Category>* marker_ca
         for (int i = 0; i < proto_pois.trail_size(); i++) {
             if (proto_pois.trail(i).map_id() == map_id) {
                 vector<string> split_categories = split(proto_pois.trail(i).category().name(), ".");
-                auto category = proto_categories.find(split_categories[split_categories.size() - 1]);
+                auto category = proto_categories.find(lowercase(split_categories[split_categories.size() - 1]));
                 if (category == proto_categories.end()) {
                     cout << "Unknown MarkerCategory " << proto_pois.trail(i).category().name() << endl;
                 }
                 else {
-                    for (string split_category : split_categories) {
-                        category_includes.insert(split_category);
+                    // This loop adds the name of the category and all of it's parents
+                    string name;
+                    for (unsigned int j = 0; j < split_categories.size(); j++) {
+                        name += split_categories[j];
+                        category_includes.insert(name);
+                        name += ".";
                     }
                 }
                 output_message.add_trail()->CopyFrom(proto_pois.trail(i));
             }
         }
+        // In XML, Marker_Categories has a tree hierarchy while POIS have a flat hierarchy.
+        // This is preserved in the protobuf for ease of translation.
+        // We are doing a removal instead of an insertion because each parent category contains the data for all of its children.
+        // It would be impractical to include all of the data from each category except for the children and then re-add the children.
+        // That would require coping every non-child attribute individually and iterating over all the children.
+        // This pruning method is slower but ensures that the all wanted information is kept.
```

Comment:
Nit: This should be styled slightly better to not exceed the 80 character column mark. There are some unclear design patterns as to when to and when not to that I can only describe as "gut feeling", but plaintext English sentences at this indentation level fall into the category of "should be line wrapped at column 80".

Also this may be the first last and only time I find a spelling mistake but "This would require coping every..." should probably be "This would require copying every..."

Reply:
Provided a suggustion.